url0307c0e995_js_labs_js_collider_tgz-pJ8-org_jsl_tests_shmem_throughput_MainJ8/src/org/jsl/collider/DatagramListenerImpl.java

```diff
@@ -0,0 +1,635 @@
+                final int newState = ST_RUNNING;
+                if (s_stateUpdater.compareAndSet(DatagramListenerImpl.this, state, ST_RUNNING))
+                {
+                    if (s_logger.isLoggable(Level.FINE))
+                    {
+                        s_logger.log( Level.FINE,
+                                m_addr + ": " + stateToString(state) + " -> " + stateToString(newState) + "." );
+                    }
+                    break;
+                }
+            }
+
+            try
+            {
+                m_selectionKey = m_datagramChannel.register(
+                        m_selector, SelectionKey.OP_READ, DatagramListenerImpl.this );
+            }
+            catch (ClosedChannelException ex)
+            {
+                /* Should not happen, means a but in the framework. */
+                if (s_logger.isLoggable(Level.WARNING))
+                    s_logger.log( Level.WARNING, ex.toString() );
+            }
+
+            return 0;
+        }
+    }
+
+    private class Starter0 extends ColliderImpl.SelectorThreadRunnable
+    {
+        public int runInSelectorThread()
+        {
+            assert( m_selectionKey.interestOps() == 0 );
+            m_selectionKey.interestOps( SelectionKey.OP_READ );
+            return 0;
+        }
+    }
+
+    private class Starter1 extends ColliderImpl.SelectorThreadRunnable
+    {
+        public int runInSelectorThread()
+        {
+            assert( m_selectionKey.interestOps() == 0 );
+            m_selectionKey.interestOps( SelectionKey.OP_READ );
+            return 1;
+        }
+    }
+
+    private static class Suspender extends ColliderImpl.SelectorThreadRunnable
+    {
+        public int runInSelectorThread()
+        {
+            return 1;
+        }
+    }
+
+    private class Stopper extends ColliderImpl.SelectorThreadRunnable
+    {
+        public int runInSelectorThread()
+        {
+            final int interestOps = m_selectionKey.interestOps();
+            if ((interestOps & SelectionKey.OP_READ) == 0)
+                m_collider.executeInSelectorThreadLater( this );
+            else
+            {
+                m_selectionKey.cancel();
+                m_selectionKey = null;
+                closeSocket();
+
+                if ((m_state & LENGTH_MASK) == 0)
+                {
+                    boolean removeListener = false;
+                    m_lock.lock();
+                    try
+                    {
+                        if (m_run)
+                        {
+                            m_run = false;
+                            m_cond.signalAll();
+                            removeListener = true;
+                        }
+                    }
+                    finally
+                    {
+                        m_lock.unlock();
+                    }
+
+                    if (removeListener)
+                        m_collider.removeDatagramListenerNoWait( m_datagramListener );
+                }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 191:210, 320:348.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.